src/core/dbus-manager.c

```diff
@@ -653,6 +653,51 @@ static int method_get_unit_by_control_group(sd_bus_message *message, void *userd
         return reply_unit_path(u, message, error);
 }
 
+static int method_get_unit_by_pidfd(sd_bus_message *message, void *userdata, sd_bus_error *error) {
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        Manager *m = ASSERT_PTR(userdata);
+        _cleanup_free_ char *path = NULL;
+        int r, pidfd;
+        pid_t pid;
+        Unit *u;
+
+        assert(message);
+
+        r = sd_bus_message_read(message, "h", &pidfd);
+        if (r < 0)
+                return r;
+
+        r = pidfd_get_pid(pidfd, &pid);
+        if (r < 0)
+                return sd_bus_error_set_errnof(error, r, "Failed to get PID from PIDFD: %m");
+
+        u = manager_get_unit_by_pid(m, pid);
+        if (!u)
+                return sd_bus_error_setf(error, BUS_ERROR_NO_UNIT_FOR_PID, "PID "PID_FMT" does not belong to any loaded unit.", pid);
+
```

Comment:
an alternative way to address my concern, if you add a `pidfd_send_signal(pidfd, 0, NULL)` check right here, so that we continue here only if the process was around during the entire time manager_get_unit_by_pid() ran.

Reply:
so, as per my latest comments, just add:

```c
r = pidfd_get_pid(pidfd, &pid2);
if (r < 0)
        return r;
if (pid != pid2)
        return sd_bus_error_set_errorf(error, BUS_ERROR_NO_SUCH_PROCESS, "Process changed PID: %m");
```
(the pid != pid2 check should never fail, except if we change pidns or so in the meantime, so that the translations change. but of course, that should be impossible, but let's better be safe here)

Reply:
(and please add a brief explanatory comment why this is needed too)